docs/isl-2-0/spec.md

```diff
@@ -0,0 +1,762 @@
+|       `[^a-cABC\d]` | complemented union codepoint class   |
+|                 `^` | anchor at the beginning of the input |
+|                 `$` | anchor at the end of the input       |
+|             `(...)` | grouping                             |
+| <code>&vert;</code> | alternation                          |
+|                 `?` | zero or one                          |
+|                 `*` | zero or more                         |
+|                 `+` | one or more                          |
+|               `{x}` | exactly x occurrences                |
+|              `{x,}` | at least x occurrences               |
+|             `{x,y}` | at least x and at most y occurrences |
+
+Regular expression flags may be specified as annotations on the regular expression string; supported flags shall include:
+
+|     |                              |
+|----:|------------------------------|
+| `i` | case insensitive             |
+| `m` | ^ and $ match at line breaks |
+
+The following classes are provided:
+
+| | |
+| `\d` | digit: `[0-9]` |
+| `\D` | non-digit |
+| `\s` | whitespace: `[ \f\n\r\t]` |
+| `\S` | non-whitespace |
+| `\w` | word character: `[A-Za-z0-9_]` |
+| `\W` | non-word character |
+
+The following characters may be escaped with a backslash:  `. ^ $ | ? * + \ [ ] ( ) { }`.
+Note that in Ion text a backslash must itself be escaped, so correct escaping of these characters requires two backslashes, e.g.:  `\\.`.
+
+
+## timestamp_offset
+
+{% include grammar-element.md productions="timestamp_offset" %}
+
+Limits the timestamp offsets that are allowed. 
+An offset is specified as a string of the form `"[+|-]hh:mm"`, where `hh` is a two-digit number between 00 and 23, inclusive, and `mm` is a two-digit number between 00 and 59, inclusive.
+The offset `"+00:00"` is equivalent to `Z`.
+The offset `"-00:00"` represents unknown local offset.
+For more details about offsets, see the [Ion Specification – Timestamps](https://amzn.github.io/ion-docs/docs/spec.html#timestamp).
+
+{% comment %}{% include example.md title="`timestamp_offset`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## timestamp_precision
+
+{% include grammar-element.md productions="timestamp_precision,timestamp_precision_value,range_timestamp_precision" %}
+
+Indicates the exact or minimum/maximum precision of a timestamp.
+Timestamp precision ranges follow the rules set out in [§Ranges](#ranges).
+Valid timestamp precision values are, in order of increasing precision: `year`, `month`, `day`, `minute`, `second`, `millisecond`, `microsecond`, and `nanosecond`.
+
+{% comment %}{% include example.md title="`timestamp_precision`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## type
+
+{% include grammar-element.md productions="type" %}
+
+Indicates the type that a value shall be validated against.
+
+{% comment %}{% include example.md title="`type`" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+## utf8_byte_length
```

Comment:
It might be a good idea to provide some guidance about when users should choose `codepoint_length` vs `utf8_byte_length`.